src/machine-learning/machine-learning.ts

```diff
@@ -229,68 +232,140 @@ export class MachineLearning implements FirebaseServiceInterface {
+    return deepCopy(this.model.tfliteModel);
+  }
+
+  /**
+   * Locked indicates if there are active long running operations on the model.
+   * Models may not be modified when they are locked.
+   */
+  public get locked(): boolean {
+    return (this.model.activeOperations?.length ?? 0) > 0;
+  }
+
+  public toJSON(): {[key: string]: any} {
+    // We can't just return this.model because it has extra fields and
+    // different formats etc. So we build the expected model object.
+    const jsonModel: {[key: string]: any}  = {
+      modelId: this.modelId,
+      displayName: this.displayName,
+      tags: this.tags,
+      createTime: this.createTime,
+      updateTime: this.updateTime,
+      published: this.published,
+      etag: this.etag,
+      locked: this.locked,
+    };
 
-    this.modelId = extractModelId(model.name);
-    this.displayName = model.displayName;
-    this.tags = model.tags || [];
-    this.createTime = new Date(model.createTime).toUTCString();
-    this.updateTime = new Date(model.updateTime).toUTCString();
-    if (model.state?.validationError?.message) {
-      this.validationError = model.state?.validationError?.message;
+    // Also add possibly undefined fields if they exist.
+
+    if (this.validationError) {
+      jsonModel['validationError'] = this.validationError;
     }
-    this.published = model.state?.published || false;
-    this.etag = model.etag;
-    if (model.modelHash) {
-      this.modelHash = model.modelHash;
+
+    if (this.modelHash) {
+      jsonModel['modelHash'] = this.modelHash;
     }
-    if (model.tfliteModel) {
-      this.tfliteModel = {
-        gcsTfliteUri: model.tfliteModel.gcsTfliteUri,
-        sizeBytes: model.tfliteModel.sizeBytes,
-      };
+
+    if (this.tfliteModel) {
+      jsonModel['tfliteModel'] = this.tfliteModel;
     }
-  }
 
-  public get locked(): boolean {
-    // Backend does not currently return locked models.
-    // This will likely change in future.
-    return false;
+    return jsonModel;
   }
 
-  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
-  public waitForUnlocked(maxTimeSeconds?: number): Promise<void> {
+
+  /**
+   * Wait for the active operations on the model to complete.
+   * @param maxTimeMillis The number of milliseconds to wait for the model to be unlocked. If unspecified,
+   *     a default will be used.
+   */
+  public waitForUnlocked(maxTimeMillis?: number): Promise<void> {
     // Backend does not currently return locked models.
     // This will likely change in future.
+    if ((this.model.activeOperations?.length ?? 0) > 0) {
+      // The client will always be defined on Models that have activeOperations
+      // because models with active operations came back from the server and
+      // were constructed with a non-empty client.
+      return this.client!.handleOperation(this.model.activeOperations![0], { wait: true, maxTimeMillis })
+        .then((modelResponse) => this.updateFromResponse(modelResponse));
```

Comment:
Nit: updateFromResponse is now a one-liner. Just `this.model = Model.validateAndClone(modelResponse);` and remove the `updateFromResponse()` helper.